### PR TITLE
Support Taproot Schnorr Signature

### PR DESCRIPTION
**Pull request by cychuang0924**
**Before submitting a pull request**, please make sure the following is done:

### Checklist

- [ ] Fork [the repository](https://github.com/getamis/alice) and create your new branch from `master`.
- [ ] Please mention the name of community in Pull Request title.
- [ ] Git Commit Messages
    - [ ] Use the present tense (Also in Pull Request title): "Add feature" not "Added feature"
    - [ ] Use the imperative mood (Also in Pull Request title): "Move cursor to..." not "Moves cursor to..."
    - [ ] Use rebase to squash/fixup dummy/unnecessary commits into only one commit.

### CODE REVIEW

File: crypto/utils/utils.go [hunk 1]
```diff
@@ -398,3 +397,12 @@ func ExtendHashOutput(salt, message []byte, outputByteLength int) []byte {
 	}
 	return result
 }
+
+func Pad(x []byte, n int) []byte {
+	pad := make([]byte, n-len(x))
```
Comment (alanchchen) [inline crypto/utils/utils.go:402]: If `len(x)` is larger than `n`, this line will panic.

File: crypto/tss/eddsa/frost/signer/signer_test.go [hunk 2]
```diff
@@ -166,3 +199,95 @@ func newSigners(curve elliptic.Curve, expPublic *ecpointgrouplaw.ECPoint, ss [][
...
+		test1, err := ecpointEncoding(R)
+		Expect(err).Should(BeNil())
+		test2 := *test1
+		r := new(big.Int).SetBytes(utils.ReverseByte(test2[:]))
+		return edwards.Verify(edwardPubKey, message, r, s)
+	}
+	return false
+}
+
+func lift_x(x *big.Int, curve elliptic.Curve) (*big.Int, *big.Int, error) {
```
Comment (alanchchen) [inline crypto/tss/eddsa/frost/signer/signer_test.go:261]: It's recommended to use camelcase. Try `liftX`

File: crypto/tss/eddsa/frost/signer/round_1.go [hunk 3]
```diff
@@ -286,61 +301,135 @@ func getMessage(messsage types.Message) *Message {
...
-		for i := 0; i < len(tempY); i++ {
-			index := len(tempY) - 1 - i
-			Y[index] = tempY[i]
+
+		return result, nil
+	}
+	return nil, ErrNotSupportCurve
+}
+
+func ecpointEncoding(pt *ecpointgrouplaw.ECPoint) (*[32]byte, error) {
```
Comment (alanchchen) [inline crypto/tss/eddsa/frost/signer/round_1.go:379]: use `[32]byte`

File: crypto/tss/eddsa/frost/signer/round_1.go [hunk 4]
```diff
@@ -245,20 +253,27 @@ func (p *round1) Finalize(logger log.Logger) (types.Handler, error) {
...
 	p.c, err = SHAPoints(p.pubKey, R, p.message)
+
 	if err != nil {
 		return nil, err
 	}
 	p.r = R
-
 	// Compute own zi = di+ ei*li + c bi xi
 	selfNode := p.nodes[p.peerManager.SelfID()]
+	if err != nil {
```
Comment (alanchchen) [inline crypto/tss/eddsa/frost/signer/round_1.go:263]: The previous line doesn't generate an error ?

File: crypto/utils/utils.go [hunk 5]
```diff
@@ -398,3 +397,12 @@ func ExtendHashOutput(salt, message []byte, outputByteLength int) []byte {
...
+	return append(pad, x...)
+}
+
+func Bytes32(x *big.Int) []byte {
```
Comment (alanchchen) [inline crypto/utils/utils.go:406]: `func Bytes32(x *big.Int) [32]byte`

File: crypto/tss/eddsa/frost/signer/round_1.go (hunk 4 above)
Comment (cychuang0924) [reply crypto/tss/eddsa/frost/signer/round_1.go:263]: done!

File: crypto/tss/eddsa/frost/signer/round_1.go (hunk 3 above)
Comment (cychuang0924) [reply crypto/tss/eddsa/frost/signer/round_1.go:379]: done

File: crypto/tss/eddsa/frost/signer/signer_test.go (hunk 2 above)
Comment (cychuang0924) [reply crypto/tss/eddsa/frost/signer/signer_test.go:261]: done

File: crypto/utils/utils.go (hunk 5 above)
Comment (cychuang0924) [reply crypto/utils/utils.go:406]: if the output is [32]byte, then we can not use "append".

File: crypto/utils/utils.go (hunk 1 above)
Comment (cychuang0924) [reply crypto/utils/utils.go:402]: Replace the implementation by:
if n-len(x) >= 0 {
	pad := make([]byte, n-len(x))
	return append(pad, x...)
}
return x[0:n]

**Pull request merged**